### PR TITLE
Added control commands and a mix of actual and stub functions

### PR DESCRIPTION
**Pull request by tails618**
* Adds (almost) all controls in `RobotContainer`
* Adds stub functions/commands for many of the controls, and simple functions/commands for others
* Adds a class for keeping track of scoring, `ScoringLocationUtil.java` 
* Added lights subsystem stub
* Added intake sequence command

### CODE REVIEW

File: src/main/java/frc/robot/RobotContainer.java [hunk 1]
```diff
@@ -105,7 +114,43 @@ public void burnFlashSparks() {
    * joysticks}.
    */
   private void configureBindings() {
-    driverController.a().whileTrue(new InstantCommand(intake::deploy, intake));
+    driverController.a().onTrue(new InstantCommand(drive::deploySkids, drive));
```
Comment (balghane) [inline src/main/java/frc/robot/RobotContainer.java]: I think this was intended to be a toggle on the diagram

File: src/main/java/frc/robot/RobotContainer.java [hunk 2]
```diff
@@ -105,7 +114,43 @@ public void burnFlashSparks() {
...
+    driverController.b().onTrue(new InstantCommand(lift::getRidOfObject, lift));
+    driverController.x().onTrue(new InstantCommand(lift::cancelScore, lift));
+    driverController.y().whileTrue(new InstantCommand(lift::manualScore, lift));
```
Comment (balghane) [inline src/main/java/frc/robot/RobotContainer.java]: Can we rename to manualPrepScore?

File: src/main/java/frc/robot/RobotContainer.java [hunk 3]
```diff
@@ -105,7 +114,43 @@ public void burnFlashSparks() {
...
+
+    driverController.back().onTrue(new InstantCommand(lift::home, lift));
```
Comment (balghane) [inline src/main/java/frc/robot/RobotContainer.java]: Can we just use go to position Starting for this?

File: src/main/java/frc/robot/RobotContainer.java [hunk 4]
```diff
@@ -105,7 +114,43 @@ public void burnFlashSparks() {
...
+    driverController.start().onTrue(new InstantCommand(drive::halfSpeedToggle, drive));
+
+    // TODO Maybe: steal
+    driverController.rightBumper().whileTrue(new InstantCommand(lift::prepScore, lift));
```
Comment (balghane) [inline src/main/java/frc/robot/RobotContainer.java]: This should probably be a press, also this will be a fully fledged command but the stub function is fine for now

File: src/main/java/frc/robot/RobotContainer.java [hunk 5]
```diff
@@ -105,7 +114,43 @@ public void burnFlashSparks() {
...
+    driverController.leftTrigger().whileTrue(new IntakeSequence(intake, lift));
+    driverController.rightTrigger().whileTrue(new Score());
```
Comment (balghane) [inline src/main/java/frc/robot/RobotContainer.java]: This should probably be a press too, and we need an onrelease or whatever that would be called

File: src/main/java/frc/robot/RobotContainer.java [hunk 6]
```diff
@@ -105,7 +114,43 @@ public void burnFlashSparks() {
...
+    driverController.b().onTrue(new InstantCommand(lift::getRidOfObject, lift));
+    driverController.x().onTrue(new InstantCommand(lift::cancelScore, lift));
+    driverController.y().whileTrue(new InstantCommand(lift::manualScore, lift));
+
+    driverController.back().onTrue(new InstantCommand(lift::home, lift));
+    driverController.start().onTrue(new InstantCommand(drive::halfSpeedToggle, drive));
+
+    // TODO Maybe: steal
+    driverController.rightBumper().whileTrue(new InstantCommand(lift::prepScore, lift));
+    driverController.leftTrigger().whileTrue(new IntakeSequence(intake, lift));
```
Comment (balghane) [inline src/main/java/frc/robot/RobotContainer.java]: Can we do onTrue and onFalse for this?

File: src/main/java/frc/robot/RobotContainer.java [hunk 7]
```diff
@@ -105,7 +114,43 @@ public void burnFlashSparks() {
...
+    driverController.rightTrigger().whileTrue(new Score());
+
+    // 0 is shelf, 1 is mid, 2 is high
```
Comment (balghane) [inline src/main/java/frc/robot/RobotContainer.java]: What are 012 here?

File: src/main/java/frc/robot/RobotContainer.java [hunk 8]
```diff
@@ -105,7 +114,43 @@ public void burnFlashSparks() {
...
+    operatorController.povDown().onTrue(new InstantCommand(() -> scoreLoc.setScoreHeight(ScoringLocationUtil.ScoreHeight.SHELF)));
+    operatorController.povLeft().onTrue(new InstantCommand(() -> scoreLoc.setScoreHeight(ScoringLocationUtil.ScoreHeight.MID)));
+    operatorController.povRight().onTrue(new InstantCommand(() -> scoreLoc.setScoreHeight(ScoringLocationUtil.ScoreHeight.MID)));
+    operatorController.povUp().onTrue(new InstantCommand(() -> scoreLoc.setScoreHeight(ScoringLocationUtil.ScoreHeight.HIGH)));
+
+    // 0 is left, 1 is center, 2 is right
```
Comment (balghane) [inline src/main/java/frc/robot/RobotContainer.java]: As before, what are 012 here?

File: src/main/java/frc/robot/RobotMap.java [hunk 9]
```diff
@@ -32,6 +32,9 @@ public final class RobotMap {
   /** Lift grabbing pneumatic channels */
   public static final int LIFT_GRABBING_FORWARD_CHANNEL = 1, LIFT_GRABBING_REVERSE_CHANNEL = 2;
 
+  /** Skid pneumatic channels */
+  public static final int SKID_FORWARD_CHANNEL = 3, SKID_REVERSE_CHANNEL = 4;
```
Comment (balghane) [inline src/main/java/frc/robot/RobotMap.java]: Let's double check, there may also be multiple skid pneumatic cylinders? We can probably just leave off all the skid details since it's so uncertain right now tho

File: src/main/java/frc/robot/commands/IntakeSequence.java [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+  }
+
+  @Override
+  public void execute() {
+    if (!lift.seeGamePiece()) {
+        intake.intakeGamePiece();
+    } else {
+        lift.grab();
+    }
+  }
```
Comment (balghane) [inline src/main/java/frc/robot/commands/IntakeSequence.java]: This is not quite right.... but we can fix it in a further PR. Can you put in a TODO?

File: src/main/java/frc/robot/subsystems/Lift.java [hunk 11]
```diff
@@ -331,6 +360,15 @@ public void periodic() {
     }
   }
 
+  /** Returns true if the lift is clear of the zone where the intake moves, so that the intake can move as soon as the lift is clear of that zone */
+  public boolean clearOfIntakeZone(){
+    if ((elevatorEncoder.getPosition() > Cal.Lift.ELEVATOR_INTAKE_ZONE_THRESHOLD_INCHES) && (armEncoder.getPosition() > Cal.Lift.ARM_INTAKE_ZONE_THRESHOLD_DEGREES)){
```
Comment (balghane) [inline src/main/java/frc/robot/subsystems/Lift.java]: I think this can be an OR

File: src/main/java/frc/robot/subsystems/Lift.java [hunk 12]
```diff
@@ -343,7 +381,7 @@ public void initSendable(SendableBuilder builder) {
     builder.addDoubleProperty("Elevator kI", elevatorPID::getI, elevatorPID::setI);
     builder.addDoubleProperty("Elevator kD", elevatorPID::getD, elevatorPID::setD);
     builder.addDoubleProperty(
-        "Elevator Position", elevatorEncoder::getPosition, elevatorEncoder::setPosition);
+        "Elevator Positionffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff", elevatorEncoder::getPosition, elevatorEncoder::setPosition);
```
Comment (balghane) [inline src/main/java/frc/robot/subsystems/Lift.java]: f

File: src/main/java/frc/robot/subsystems/drive/DriveSubsystem.java [hunk 13]
```diff
@@ -68,6 +70,10 @@ public class DriveSubsystem extends SubsystemBase {
             rearRight.getPosition()
           });
 
+  private boolean halfSpeed = false;
```
Comment (balghane) [inline src/main/java/frc/robot/subsystems/drive/DriveSubsystem.java]: Add a TODO to actually use this?

File: src/main/java/frc/robot/utils/ScoringLocationUtil.java [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+  }
+
+  /** Column to score in */
+  public enum ScoreCol {
+    LEFT,
+    CENTER,
+    RIGHT
+  }
+
+  private ScoreHeight scoreHeight = ScoreHeight.SHELF; // default value
```
Comment (balghane) [inline src/main/java/frc/robot/utils/ScoringLocationUtil.java]: Super minor but let's default to HIGH

File: src/main/java/frc/robot/utils/ScoringLocationUtil.java [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+    middleGrid = middleGrid ? false : true;
+  }
+
+  public void setScoreCol(ScoreCol chooseCol) {
+    scoreCol = chooseCol;
+  }
+
+  public void setScoreHeight(ScoreHeight chooseHeight) {
+    scoreHeight = chooseHeight;
+  }
```
Comment (balghane) [inline src/main/java/frc/robot/utils/ScoringLocationUtil.java]: Add getters as well. Add a TODO to map these to april tag IDs

**Review by balghane — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/commands/IntakeSequence.java [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+    return lift.holdingGamePiece() ? true : false;
+  }
+
+  @Override
+  public void end(boolean interrupted){
+    if (!interrupted){
+      lift.setDesiredPosition(Lift.LiftPosition.STARTING);
+      if (lift.clearOfIntakeZone()){
+        intake.retract();
+      }
```
Comment (balghane) [inline src/main/java/frc/robot/commands/IntakeSequence.java]: We should just put this logic right in the intake, let it query the lift. But we can do that in a separate PR.

File: src/main/java/frc/robot/RobotContainer.java (hunk 3 above)
Comment (tails618) [reply src/main/java/frc/robot/RobotContainer.java]: `goToPosition` is private, so I can't access it from here.

**Comment by balghane** [comment]
setDesired?

File: src/main/java/frc/robot/RobotContainer.java (hunk 5 above)
Comment (tails618) [reply src/main/java/frc/robot/RobotContainer.java]: What would the `onFalse` do? End the command?

File: src/main/java/frc/robot/subsystems/Lift.java (hunk 12 above)
Comment (tails618) [reply src/main/java/frc/robot/subsystems/Lift.java]: oops

Comment (tails618) [reply src/main/java/frc/robot/subsystems/Lift.java]: i blame dell for this

**Review by balghane — APPROVED** [review]
